src/macaron/repo_finder/commit_finder.py

```diff
@@ -0,0 +1,397 @@
+
+    Parameters
+    ----------
+    version: str
+        The version string.
+
+    Returns
+    -------
+    tuple[Pattern, list[str], bool]
+        The tuple of the regex pattern that will match the version, the list of version parts that were extracted, and
+        whether the version string has a non-numeric suffix.
+
+    """
+    # The version is split on non-alphanumeric characters to separate the version parts from the non-version parts.
+    # e.g. 1.2.3-DEV -> [1, 2, 3, DEV]
+    split = split_pattern.split(version)
+    logger.debug("Split version: %s", split)
+    if not split:
+        # If the version string contains no separators use it as is.
+        split = [version]
+
+    this_version_pattern = ""
+    parts = []
+    numeric_parts = 0
+    has_non_numeric_suffix = False
+    for count, part in enumerate(split):
+        # Validate the split part by checking it is only comprised of alphanumeric characters.
+        valid = validation_pattern.match(part)
+        if not valid:
+            continue
+        parts.append(part)
+
+        # Detect versions that end with a zero, so the zero can be made optional.
+        has_trailing_zero = len(split) > 2 and count == len(split) - 1 and part == "0"
+        numeric_only = numeric_only_pattern.match(str(part))
+
+        if not has_non_numeric_suffix and not numeric_only:
+            # A non-numeric part enables the flag for treating this and all remaining parts as version suffix parts.
+            has_non_numeric_suffix = True
+
+        if has_trailing_zero or has_non_numeric_suffix:
+            # This part will be made optional in the regex, hence the grouping bracket.
+            this_version_pattern = this_version_pattern + "("
+
+        if not has_non_numeric_suffix and numeric_only:
+            # Keep track of version length for upcoming possible 'zero' digit extensions.
+            numeric_parts = numeric_parts + 1
+
+        if count == 1:
+            this_version_pattern = this_version_pattern + INFIX_1
+        elif count > 1:
+            this_version_pattern = this_version_pattern + INFIX_3
+
+        # Add the current part to the pattern.
+        this_version_pattern = this_version_pattern + str(part)
+
+        if has_trailing_zero or has_non_numeric_suffix:
+            # Complete the optional capture group.
+            this_version_pattern = this_version_pattern + ")?"
+
+        # Extend the pattern with up to three additional zeros to help map mismatches between versions and tags.
+        # E.g. 1.0 to 1.0.0, or 3.2 to 3.2.0, etc.
+        if not has_non_numeric_suffix and count == len(split) - 1 and 1 < numeric_parts < 4:
+            while numeric_parts < 4:
+                # Additional zeros added for this purpose make use of a back reference to the first matched separator.
+                this_version_pattern = this_version_pattern + "(" + INFIX_2 + "0)?"
+                numeric_parts = numeric_parts + 1
```

Comment:
This `if` statement only runs when the outer `for` loop reaches the last iteration, and it literally runs at the end of this last iteration. Therefore I think it should be moved out of the `for` loop.